Document new 'config' API (#4331)
tarantool/doc: doc/reference/reference_lua/config.rst

    .. _config_api_reference_get:

    ..  method:: get([param, opts])

        Get a configuration applied to the current or remote instance.

>         Note that there is the following difference between getting a configuration for the current and remote instance:

Reviewer: ```suggestion
        Note the following differences between getting a configuration for the current and remote instance:
```

>         -   For the current instance, ``get()`` returns an instance configuration considering :ref:`environment variables <configuration_environment_variable>`.

Reviewer: ```suggestion
        -   For the current instance, ``get()`` returns instance configuration considering :ref:`environment variables <configuration_environment_variable>`.
```

>         -   For a remote instance, ``get()`` only considers a cluster configuration and ignores environment variables.

Reviewer: ```suggestion
        -   For a remote instance, ``get()`` only considers cluster configuration and ignores environment variables.
```

    .. _config_api_reference_info:

>     ..  method:: info([version])

Reviewer 2: I would update the examples below to use `v2` format. It is more explicit and I would recommend users to use it.

The `v2` format is also used in `box.info.config` (see #4238).

Reviewer 2: Nice, thanks!

                 -   ``status`` -- one of the following statuses:

                     *   ``ready`` -- the configuration is applied successfully
                     *   ``check_warnings`` -- the configuration is applied with warnings
                     *   ``check_errors`` -- the configuration cannot be applied due to configuration errors

                 -   ``meta`` -- additional configuration information

                 -   ``alerts`` -- warnings or errors raised on an attempt to apply the configuration

        Below are a few examples demonstrating how the ``info()`` output might look.

        **Example: no configuration warnings or errors**

        In the example below, an instance's state is ``ready`` and no warnings are shown:

        ..  code-block:: tarantoolsession

            app:instance001> require('config'):info('v2')
            ---
            - status: ready
              meta:
                last: &0 []
                active: *0
              alerts: []
            ...

        **Example: configuration warnings**

        In the example below, the instance's state is ``check_warnings``.

>         The ``alerts`` section informs that privileges to the ``bands`` space for ``sampleuser`` cannot be granted because the ``bands`` space has not created yet:

Reviewer: ```suggestion
        The ``alerts`` section informs that privileges to the ``bands`` space for ``sampleuser`` cannot be granted because the ``bands`` space has not been created yet:
```

        ..  code-block:: tarantoolsession

            app:instance001> require('config'):info('v2')
            ---
            - status: check_errors
              meta:
                last:
                  etcd:
                    mod_revision:
                      /myapp/config/all: 5
                    revision: 5
                active:
                  etcd:
                    mod_revision:
                      /myapp/config/all: 2
                    revision: 4
              alerts:
              - type: error
                message: 'etcd source: invalid config at key "/myapp/config/all": [cluster_config]
                  groups.group001.replicasets.replicaset001.instances.instance001.log.level: Got
                  8, but only the following values are allowed: 0, fatal, 1, syserror, 2, error,
                  3, crit, 4, warn, 5, info, 6, verbose, 7, debug'
                timestamp: 2024-07-03T15:22:06.438275Z
            ...

    .. _config_api_reference_instance_uri:

    ..  method:: instance_uri([uri_type, opts])

        **Since:** :doc:`3.1.0 </release/3.1.0>`

        Get a URI of the current or remote instance.

>         :param string uri_type: a URI type. There are the following URI types are supported:

Reviewer: ```suggestion
        :param string uri_type: a URI type. The following URI types are supported:
```

        List all instances of the cluster.

        :return: a table containing information about instances. The returned table uses instance names as the keys and contains the following information for each instance:

                 -   ``instance_name`` -- an instance name
                 -   ``replicaset_name`` -- the name of a replica set the instance belongs to
                 -   ``group_name`` -- the name of a group the instance belongs to

        **Example**

        The example below shows how to use ``instances()`` to get the names of all instances in the cluster, create a connection to each instance using the :ref:`connpool <connpool_module>` module, and log connection URIs using the :ref:`log <log-module>` module:

        ..  code-block:: lua

            local config = require('config')
            local connpool = require('experimental.connpool')
            local log = require('log')

            for instance_name in pairs(config:instances()) do
                local conn = connpool.connect(instance_name)
                log.info("Connection URI for %q: %s:%s", instance_name, conn.host, conn.port)
            end

>         In this example, the same actions are performed for instances from the specific replica set:

Reviewer: ```suggestion
        In this example, the same actions are performed for instances from a specific replica set:
```

                  alerts: []
                ...

    .. _config_api_reference_instance_uri:

    ..  method:: instance_uri([uri_type, opts])

        **Since:** :doc:`3.1.0 </release/3.1.0>`

        Get a URI of the current or remote instance.

        :param string uri_type: a URI type. There are the following URI types are supported:

                                * ``peer`` -- a URI used to advertise the instance to other cluster members. See also: :ref:`iproto.advertise.peer <configuration_reference_iproto_advertise_peer>`.
                                * ``sharding`` -- a URI used to advertise the current instance to a router and rebalancer. See also: :ref:`iproto.advertise.sharding <configuration_reference_iproto_advertise_sharding>`.

        :param table opts: options to pass. The following options are available:

                           -   ``instance`` -- the name of a remote instance whose URI should be obtained

>         :return: an instance URI

Reviewer 2: I agree that all the details how the URI is constructed are redundant here. I think I should create a section in the internals and link it from here. It's on me.

But I would add information about the fields in the returned table right into the reference: it is part of the method contract. Also, it worth to add a note that the return value is suitable to pass into `net_box.connect()` (otherwise it may look logical to extract the `uri` field from it before calling `net_box.connect()`).

            config:instance_uri('peer', { instance = 'storage-b-003' })

    .. _config_api_reference_instances:

    ..  method:: instances()

        **Since:** :doc:`3.1.0 </release/3.1.0>`

        List all instances of the cluster.

        :return: a table containing information about instances

        The returned table uses instance names as the keys and contains the following information for each instance:

        -   ``instance_name`` -- an instance name
        -   ``replicaset_name`` -- the name of a replica set the instance belongs to
        -   ``group_name`` -- the name of a group the instance belongs to

        **Example**

        The example below shows how to use ``instances()`` to get the names of all instances in the cluster, create a connection to each instance using the :ref:`connpool <connpool_module>` module, and log connection URIs using the :ref:`log <log-module>` module:

        ..  code-block:: lua

            local config = require('config')
            for instance_name in pairs(config:instances()) do
                local connpool = require('experimental.connpool')
                local conn = connpool.connect(instance_name)

>                 local log = require('log')

Reviewer 2: Stylistic suggestion: I would move `require` calls to the top.

>                 log.info(string.format("Connection URI for '%s': %s:%s", instance_name, conn.host, conn.port))

Reviewer 2: BTW, `%q` can be used instead of `"%s"` for quoting.

```lua
tarantool> log.info('xxx %q zzz', 'yyy')
xxx "yyy" zzz
```

And an explicit `string.format` is not needed, `log.info` calls it itself.

>             end

Reviewer 2: Maybe it wotrh to add some example that uses the second iteration value: `for instance_name, def in pairs(config:instances()) do <...> end`?

#4105 has an example with filtering instances of the same replicaset.
